src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/IHubProtocol.cs

```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Sockets;
+
+namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
+{
+    public interface IHubProtocol
+    {
+        MessageType MessageType { get; }
+
+        bool TryParseMessage(ReadOnlySpan<byte> input, IInvocationBinder binder, out HubMessage message);
+
+        // Need a better API when we have sorted out pooling, etc.
+        // See https://github.com/aspnet/SignalR/issues/126
+        // For exmaple: bool TryWriteMessage(HubMessage message, IOutput output);
+        byte[] WriteMessage(HubMessage message);
```

Comment:
It would have to be `IOutput`, otherwise we have to handle allocation strategies everywhere it's called. I started with that, but the JSON serialization is all `Stream`-based, so it ends up just using `MemoryStream` and `ToArray` to get the data, then it just bulk-writes that to the `IOutput`.

And then, we have to build our own `IOutput` implementation over a buffer pool, or allocate a pipeline. I didn't feel like we were ready for that so I used this API rather than try to shoe-horn a messy implementation pattern into the "desired" API. We need a proper plan for how to actually handle pooling of memory. If we want to block this change on getting that figured out, I can work on it, but I didn't want to overload this PR.